server/blob/mail-store/src/main/java/org/apache/james/blob/mail/MimeMessageStore.java

```diff
@@ -0,0 +1,144 @@
+            } catch (MessagingException | IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        private static byte[] messageToArray(MimeMessage message) throws IOException, MessagingException {
+            ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+            message.writeTo(byteArrayOutputStream);
+            return byteArrayOutputStream.toByteArray();
+        }
+
+        private static byte[] getHeaderBytes(byte[] messageContentAsArray, int bodyStartOctet) {
+            ByteBuffer headerContent = ByteBuffer.wrap(messageContentAsArray, 0, bodyStartOctet);
+            byte[] headerBytes = new byte[bodyStartOctet];
+            headerContent.get(headerBytes);
+            return headerBytes;
+        }
+
+        private static byte[] getBodyBytes(byte[] messageContentAsArray, int bodyStartOctet) {
+            if (bodyStartOctet < messageContentAsArray.length) {
+                ByteBuffer bodyContent = ByteBuffer.wrap(messageContentAsArray,
+                    bodyStartOctet,
+                    messageContentAsArray.length - bodyStartOctet);
+                byte[] bodyBytes = new byte[messageContentAsArray.length - bodyStartOctet];
+                bodyContent.get(bodyBytes);
+                return bodyBytes;
+            } else {
+                return new byte[] {};
+            }
+        }
+
+        private static int computeBodyStartOctet(byte[] messageAsArray) throws IOException {
+            try (BodyOffsetInputStream bodyOffsetInputStream =
+                     new BodyOffsetInputStream(new ByteArrayInputStream(messageAsArray))) {
+                consume(bodyOffsetInputStream);
+
+                if (bodyOffsetInputStream.getBodyStartOffset() == -1) {
+                    return 0;
+                }
+                return (int) bodyOffsetInputStream.getBodyStartOffset();
+            }
+        }
+
+        private static void consume(InputStream in) throws IOException {
+            IOUtils.copy(in, NULL_OUTPUT_STREAM);
+        }
+    }
+
+    static class MailDecoder implements Decoder<MimeMessage> {
+        @Override
+        public void validateInput(Collection<BlobType> input) {
+            Preconditions.checkArgument(input.contains(HEADER_BLOB_TYPE), "Expecting 'mailHeader' blobId to be specified");
+            Preconditions.checkArgument(input.contains(BODY_BLOB_TYPE), "Expecting 'mailBody' blobId to be specified");
+            Preconditions.checkArgument(input.size() == 2, "blobId other than 'mailHeader' or 'mailBody' are not supported");
+        }
+
+        @Override
+        public MimeMessage decode(Stream<Pair<BlobType, byte[]>> streams) {
+            Preconditions.checkNotNull(streams);
+            Map<BlobType,byte[]> pairs = streams.collect(ImmutableMap.toImmutableMap(Pair::getLeft, Pair::getRight));
+            Preconditions.checkArgument(pairs.containsKey(HEADER_BLOB_TYPE));
```

Comment:
This statement is duplicated in `validateInput` method, looks like a validating action inside `decode`. How about calling `validateInput` inside `decode`?

Reply:
This problem is IMO creatred by the intermediate use of Stream, what I stood against :-)

Reply:
I don't get it, but you can user `validateInput(pairs.keySet());` instead of the current statement

Reply:
Validate is moved into the id factory